File: LeetCode/src/problem301_400/Problem396_1.java
package problem301_400;

public class Problem396_1 {

    public int maxRotateFunction(int[] nums) {
        final int n = nums.length;
        int res = 0, last = 0, sum = 0;
        for (int i = 0; i < n; ++i) {
            res += i * nums[i];
            sum += nums[i];
        }
        last = res;

        for (int i = n - 1; i > 0; --i) {
            last += (sum - n * nums[i]);
            if (res < last) {
                res = last;
            }
        }

        return res;
    }

}

File: LeetCode/src/problem301_400/Problem396_1.kt
package problem301_400

import kotlin.math.max

fun maxRotateFunction(nums: IntArray): Int {
    val sum = nums.sum()
    var ansMax = 0
    nums.forEachIndexed { index, value ->
        ansMax += index * value
    }
    var preValue = ansMax
    val n = nums.size
    for (k in 1 until n) {
        preValue += sum - n * nums[n - k]
        ansMax = max(ansMax, preValue)
    }
    return ansMax
}
